crates/libcontainer/src/container/container_kill.rs

```diff
@@ -24,37 +24,87 @@ impl Container {
     /// # }
     /// ```
     pub fn kill<S: Into<Signal>>(&mut self, signal: S, all: bool) -> Result<()> {
-        let signal = signal.into().into_raw();
-
-        let pids = if all {
-            let cgroups_path = self.spec()?.cgroup_path;
-            let use_systemd = self
-                .systemd()
-                .context("container state does not contain cgroup manager")?;
-            let cmanger = create_cgroup_manager(&cgroups_path, use_systemd, self.id())?;
-            cmanger.get_all_pids()?
-        } else {
-            vec![self
-                .pid()
-                .context("failed to get the pid of the container")?]
-        };
-
         self.refresh_status()
             .context("failed to refresh container status")?;
         if self.can_kill() {
-            pids.into_iter().try_for_each(|pid| {
-                log::debug!("kill signal {} to {}", signal, pid);
-                signal::kill(pid, signal)
-            })?;
+            self.do_kill(signal, all)?;
+        } else {
+            // just like runc, allow kill --all even if the container is stopped
+            if all && self.status() == ContainerStatus::Stopped {
```

Comment:
Shouldn't here we check for stopped and running both?  Otherwise it will give error when using --all with running container.
Also, given that we are having an if-else in the `do_kill` function, would it be better to move the checking logic there, or remove that function completely and move its logic here?

Reply:
https://github.com/containers/youki/blob/23170c6a3225d951fe556ac652ccd0c5df3a9c68/crates/libcontainer/src/container/state.rs#L40-L46
https://github.com/containers/youki/blob/d4088a9a61f869d951739eb7b5bc5ccc23691292/crates/libcontainer/src/container/container_kill.rs#L29

If container is running, it will pass the `can_kill` check and be killed without error. Maybe youki should be the same as runc, remove the check for `Stopped` status here, and do the kill in all status if with the --all option.

I think `do_fork` should not add checking logic. In this way, it can be used more flexibly by other modules, such as  `delete` command.

Reply:
Hey, I agree with you, this makes sense :+1: Let us keep it the way it is, I'll try to open an issue on runc to see what can be done.